### ISSUE TITLE
[Feature Request]: Command Rewards should use the player in their context even when not run as the player

### ISSUE THREAD

**Issue by FunnyMan3595** [FTBTeam/FTB-Mods-Issues#716]
### Mod

FTB Quests

### Description

The command reward currently has two modes, based on the "Run as Player" setting.  With the setting on, the command executes exactly as if the player had typed it themselves.  With the setting off, the command acts as if it were entered into the server console.

Relevant code:
https://github.com/FTBTeam/FTB-Quests/blob/1.19/main/common/src/main/java/dev/ftb/mods/ftbquests/quest/reward/CommandReward.java#L98

The problem with this is that if you want to run a command that the player can't normally run, but try to reference the player with `@p`, it will just pick an arbitrary player, not the one who completed the quest.

### Describe the solution you'd like

Instead of using a default command stack when the setting is off, use the player's command stack anyway, but add .withPermission(2) to it.  This creates an operator-level command stack, regardless of the player's permission level, but still maintains the information about the player for `@p`, `~`, and similar.

### Describe alternatives you've considered

You could also make it a third way to run the command, separate from both of the existing ones, but I don't see any particular reason that it would be harmful to attach the player's data to all commands.

**Comment by desht** [comment]
I agree, this does feel like a better way to do it. It will need a little consideration though, and while I also can't offhand see a problem with changing existing behaviour, changes like this do have a tendency to cause surprises (e.g. are there commands that really must be run as server console commands...)

**Comment by desht** [comment]
Actually, looking again at the `CommandReward` class, there is some specific substitution going on in the command; in particular `@p` is substituted with the name of the player receiving the reward.  So you shouldn't be getting an arbitrary player there.

Is it your experience that this isn't working?

**Comment by FunnyMan3595** [comment]
Aha, that'd be partially user error, then.  The command I was using doesn't take any arguments at the moment, and was re-using the command sender to dispatch sub-commands, but I can change it to accept one and work around this that way.

Still a strange way to do things, though.

**Comment by desht** [comment]
Yeah, it would be best to use Minecraft's own substitution system here, I agree.  A lot of this is quite historical, and due for a major rewrite for MC 1.20+.

In the short term though, I'm wary of breaking any existing setups, so your option 3 - add a "elevated permission" option - is probably the way to go for now.

**Comment by SaintObstinate** [comment]
Any news on this thread? 
I'm trying to make weekly quests that resets every sunday, I looked around the internet and I came up with only two solutions:
1- Automate a command on server to send _/ftbquests change_progress_ to all players. But the problem with this, is that it only works for the online players, if no one is online, the command wont work (is that a way for this to work offline?)
2- Add a timer quest (with the quest additions mod) and the reward is the command one, but the problem is the same of the OP on this thread.

**Comment by desht** [comment]
This is done in FTB Quests 1.20, which has a "Run with Elevated Permission" option in the reward properties - if true, the command is run with permission level 2.

There isn't such an option in 1.19 though - not sure which version you're currently on?

**Comment by Compl3xlife** [comment]
> ### Mod
> FTB Quests
> 
> ### Description
> The command reward currently has two modes, based on the "Run as Player" setting. With the setting on, the command executes exactly as if the player had typed it themselves. With the setting off, the command acts as if it were entered into the server console.
> 
> Relevant code: https://github.com/FTBTeam/FTB-Quests/blob/1.19/main/common/src/main/java/dev/ftb/mods/ftbquests/quest/reward/CommandReward.java#L98
> 
> The problem with this is that if you want to run a command that the player can't normally run, but try to reference the player with `@p`, it will just pick an arbitrary player, not the one who completed the quest.
> 
> ### Describe the solution you'd like
> Instead of using a default command stack when the setting is off, use the player's command stack anyway, but add .withPermission(2) to it. This creates an operator-level command stack, regardless of the player's permission level, but still maintains the information about the player for `@p`, `~`, and similar.
> 
> ### Describe alternatives you've considered
> You could also make it a third way to run the command, separate from both of the existing ones, but I don't see any particular reason that it would be harmful to attach the player's data to all commands.

how would that command look? to add permission(2) i mean.

**Comment by manmaed** [comment]
Please disregard the previous reply from the stale bot, it's been a bad bot and sent for recycling.
!skip

**Comment by MichaelHillcox** [comment]
This is sorted. This won't be backported from 1.20 downwards
